### PR TITLE
Move the error from TrieValue to CodePointTrie

### LINKED ISSUE #1879: Move error_value (and optionally default_value) to the CPT header?
See previous discussion in #1183.

The thread #1183 got derailed into a discussion on trust boundaries. I moved that conversation over to #1290. Our work on CrabBake is our solution to that side of the problem and allows us to reason about "invalid serialized data" in the serde case.

I increasingly feel that the option most consistent with how our data model works is to move these two fields, especially `error_value`, into the CPT header. This would help us eliminate the problematic `DATA_GET_ERROR_VALUE`.

This should be an easy change: the error value is always the last value in the data array, so at datagen time, we just remove the last value from the array and save it into the header.

Responses to the pushback on this idea from the previous thread:

> @markusicu: Between UTrie2 and the newer CodePointTrie, I moved the error value from the header struct into the data array in order to make the code simpler and more uniform.

In Rust, I do not believe that this will increase code complexity. It would involve adding a type parameter to `CodePointTrieHeader`, but this parameter would just be bubbled down from `CodePointTrie`, which already has a type parameter.

It may require adding `#[derive(Deserialize)]` on TrieType types that don't already have it.

> @hsivonen: This would make ICU4X compute bogus results if externally-traveled data actually has been corrupted. In terms of surprise and what it takes to understand what's going wrong, this seems worse than any of the other options. However, in terms of usage ergonomics, I'd rather settle on this option than option 2 if we can't agree on panicking on malformed data at the time of making lookups from the data.

We are moving from one type of GIGO to another type of GIGO. The new type of GIGO is in fact slightly more robust, because it allows us to automatically fail in the constructor if the error value field is not present in the header, something we can't as easily do when it is stored in the data array. Discussion of when or when not to rely on GIGO is the topic of #1290 and I do not want to derail this thread again on that separate, very important issue.

Needs approval from:

- [x] @echeran 
- [ ] @markusicu 
- [x] @Manishearth

### PR DESCRIPTION
**Pull request by pandusonu2**
Fixes #1879

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]

File: utils/codepointtrie/src/codepointtrie.rs [hunk 1]
```diff
@@ -390,6 +395,7 @@ impl<'trie, T: TrieValue> CodePointTrie<'trie, T> {
             header: self.header,
             index: self.index,
             data: converted_data,
+            error_value: self.error_value,
```
Comment (Manishearth) [inline utils/codepointtrie/src/codepointtrie.rs:398]: hm, @sffc how should try_into_converted work here? should the error value also just be casted with a try_into?

Comment (Manishearth) [reply utils/codepointtrie/src/codepointtrie.rs:398]: The way to do this with ZeroVec would be:

 - Convert `self.error_value` to `T::ULE` by calling `self.error_value.to_unaligned()`
 - Construct a `ZeroVec::Borrowed(&[ule_value])`
 - Call `try_into_converted()` on _that_
 - Call `.get(0)` on that and use that value

Comment (sffc) [reply utils/codepointtrie/src/codepointtrie.rs:398]: Hmm. Hadn't considered that. @Manishearth's workaround sounds good to me.

**Comment by sffc** [comment]
@pandusonu2 When do you plan to finish this PR?

**Comment by Manishearth** [comment]
Superseded by  #2301

**Pull request closed without merging**